### PR TITLE
Add icu_provider_fs feature to datagen; bump to 1.2.2

### PR DESCRIPTION
**Pull request by Manishearth**
Not sure what the feature itself should be called.

Most users of datagen shouldn't actually need the fs feature, and it pulls in some deps + enables the `sync` feature on `icu_provider`, which we should avoid in unified-build situations.

### CODE REVIEW

File: provider/datagen/Cargo.toml [hunk 1]
```diff
@@ -106,6 +106,7 @@ use_wasm = ["icu_codepointtrie_builder/wasm"]
 # rule based segmenter data will not be generated.
 use_icu4c = ["icu_codepointtrie_builder/icu4c"]
 networking = ["dep:cached-path"]
+icu_provider_fs = ["dep:icu_provider_fs"]
```
Comment (sffc) [inline provider/datagen/Cargo.toml:109]: Thought: maybe the feature should be named `fs`? The feature isn't really about the dependency; it's about the functionality, and the dependency is a convenient side-effect.

File: provider/datagen/src/bin/datagen.rs [hunk 2]
```diff
@@ -333,6 +335,9 @@ fn main() -> eyre::Result<()> {
             if v == cli::Format::DeprecatedDefault {
                 log::warn!("Defaulting to --format=dir. This will become a required parameter in the future.");
             }
+            #[cfg(not(feature = "icu_provider_fs"))]
+            eyre::bail!("--format=dir only available with the icu_provider_fs feature");
```
Comment (sffc) [inline provider/datagen/src/bin/datagen.rs:339]:
```suggestion
            eyre::bail!("--format=dir only available with the icu_provider_fs Cargo feature");
```

File: provider/datagen/src/lib.rs [hunk 3]
```diff
@@ -315,6 +319,9 @@ impl Default for BakedOptions {
 #[non_exhaustive]
 pub enum Out {
     /// Output to a file system tree
+    ///
+    /// Only available with the `icu_provider_fs` feature
```
Comment (sffc) [inline provider/datagen/src/lib.rs:323]: Nit:

```suggestion
    /// **Enabled with the `icu_provider_fs` Cargo feature.**
```

As discussed previously, a large percentage of problems our users have is when they don't have the right features enabled, so I think we should try to continue to be loud and consistent on functions that need a Cargo feature.

File: provider/datagen/Cargo.toml (hunk 1 above)
Comment (Manishearth) [reply provider/datagen/Cargo.toml:109]: Yeah `icu_provider_fs` was a placeholer. I like `fs`, was also thinking `filesystem` to match `networking` but they're talking about different kinds of things

File: provider/datagen/Cargo.toml [hunk 4]
```diff
@@ -5,7 +5,7 @@
 [package]
 name = "icu_datagen"
 description = "Generate data for ICU4X DataProvider"
-version = "1.2.1"
+version = "1.2.2"
```
Comment (sffc) [inline provider/datagen/Cargo.toml:8]: Nit: Update the changelog (can be a separate PR). Also the lockfiles.

**Review by sffc — APPROVED** [review]

**Review by sffc — APPROVED** [review]

File: provider/datagen/Cargo.toml [hunk 5]
```diff
@@ -106,6 +106,7 @@ use_wasm = ["icu_codepointtrie_builder/wasm"]
 # rule based segmenter data will not be generated.
 use_icu4c = ["icu_codepointtrie_builder/icu4c"]
 networking = ["dep:cached-path"]
+fs = ["dep:icu_provider_fs"]
```
Comment (robertbastian) [inline provider/datagen/Cargo.toml:109]: This should be a default feature

File: provider/datagen/src/bin/datagen.rs [hunk 6]
```diff
@@ -333,6 +335,9 @@ fn main() -> eyre::Result<()> {
             if v == cli::Format::DeprecatedDefault {
                 log::warn!("Defaulting to --format=dir. This will become a required parameter in the future.");
             }
+            #[cfg(not(feature = "fs"))]
+            eyre::bail!("--format=dir only available with the `fs` Cargo feature");
```
Comment (robertbastian) [inline provider/datagen/src/bin/datagen.rs:339]: `fs` is not a default feature so this is a breaking change

File: provider/datagen/src/bin/datagen.rs [hunk 7]
```diff
@@ -333,6 +335,9 @@ fn main() -> eyre::Result<()> {
             if v == cli::Format::DeprecatedDefault {
                 log::warn!("Defaulting to --format=dir. This will become a required parameter in the future.");
```
Comment (robertbastian) [inline provider/datagen/src/bin/datagen.rs:336]: I don't think it makes sense to default to something that will panic

File: provider/datagen/Cargo.toml (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/Cargo.toml]: I'd prefer `provider_fs` over `fs`. Datagen still makes extensive use of the filesystem.

**Review by robertbastian — COMMENTED** [review]
Was there a pressing reason for merging and releasing this outside european working hours?

**Comment by robertbastian** [comment]
> it pulls in some deps + enables the sync feature on icu_provider, which we should avoid in unified-build situations.

For the record it pulls in a single dep, `serde-json-core`, and does not enable the sync feature (icu_provider/datagen does that as it's needed for datagen parallelism).

File: provider/datagen/src/lib.rs [hunk 8]
```diff
@@ -315,6 +319,9 @@ impl Default for BakedOptions {
 #[non_exhaustive]
 pub enum Out {
     /// Output to a file system tree
+    ///
+    /// **Enabled with the `fs` Cargo feature.**
+    #[cfg(feature = "fs")]
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply provider/datagen/src/lib.rs:324]: Oops, it was in my original local draft and I think i lost it
